odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+
+    Given a functional ``f`` and a scalar ``scalar``, this represents the
+    functional
+
+        ``(scalar * f)(x) = scalar * f(x)``.
+    """
+
+    def __init__(self, func, scalar):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The functional that is to be scaled.
+        scalar : `float`
+            The scalar with which the functional is scaled.
+        """
+
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is not a Functional instance.'
+                            ''.format(func))
+
+        scalar = func.range.element(scalar)
+
+        if scalar > 0 or scalar < 0:
+            Functional.__init__(self, domain=func.domain,
+                                linear=func.is_linear,
+                                grad_lipschitz=(
+                                    np.abs(scalar) * func.grad_lipschitz))
+
+        elif scalar == 0:
+            Functional.__init__(self, domain=func.domain, linear=True,
+                                grad_lipschitz=0)
+
+        else:
+            # It should not be possible to get here
+            raise TypeError('comparison with scalar {} failed'.format(scalar))
+
+        OperatorLeftScalarMult.__init__(self, operator=func, scalar=scalar)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return self.scalar * self.operator.gradient
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the scaled functional."""
+        # The helper function only allows positive scaling parameters.
+        # Otherwise it gives an error.
+        return ConvexConjugateFuncScaling(
+            self.operator.conjugate_functional, self.scalar)
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the scaled functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+        """
+
+        sigma = float(sigma)
+
+        if self.scalar < 0:
+            raise ValueError('Proximal operator of functional scaled with a '
+                             'negative value {} is not well-defined'
+                             ''.format(self.scalar))
+
+        elif self.scalar == 0:
+            return IdentityOperator(self.domain)
+
+        elif self.scalar > 0:
+            return self.operator.proximal(sigma * self.scalar)
```

Comment:
The thing is that `np.inf` will pass all checks and be passed to `proximal`. That may be okay since some proximal factory implementation may decide to allow this. We just need to keep in mind that it may happen and needs to be handled.